### ISSUE TITLE
[feature request] GPLVMs with missing data

### ISSUE THREAD

**Issue by InfProbSciX** [pyro-ppl/pyro#2654]
I'm interested in using the `pyro.contrib.gp.models.GPLVM` model in the case where some points in `Y` (the observed data) are missing at random.

GPy has an example [here](https://gpy.readthedocs.io/en/deploy/_modules/GPy/examples/dimensionality_reduction.html#bgplvm_simulation_missing_data).

This functionality isn't available in pyro at the moment (I think - couldn't find examples for this). Is this something that can be added? I'd be happy to contribute some code to do this.

**Comment by fehiepsi** [comment]
Hi @InfProbSciX, welcome! Feel free to make a PR for that feature. I guess to deal with NaN, we apply mask to the likelihood?

**Comment by InfProbSciX** [comment]
@fehiepsi Yep, I do believe so. Here's a quick snippet where I apply a masked likelihood, but it doesn't work unfortunately. The gradients computed are `nan`s, so after the first optimize step, training breaks.

Any ideas on how to fix this?

```python
import pyro
import torch
import unittest
import numpy as np
import pyro.contrib.gp as gp
import pyro.distributions as dist
from pyro.infer.util import torch_backward

SEED = 42
torch.manual_seed(SEED); pyro.set_rng_seed(SEED); np.random.seed(SEED)

class MaskedGaussian(gp.likelihoods.Gaussian):
    def forward(self, f_loc, f_var, y=None):
        y_var = f_var + self.variance
        y_dist = dist.Normal(f_loc, y_var.sqrt())

        if y is not None:
            if y.isnan().any():
                y_dist = dist.MaskedDistribution(y_dist, ~y.isnan())

            y_dist = y_dist.expand_by(y.shape[:-f_loc.dim()]).to_event(y.dim())
        return pyro.sample(self._pyro_get_fullname("y"), y_dist, obs=y)

n, m, q = 10, 3, 2
Y = torch.tensor(np.random.normal(size=(n, q))).float()
Y[0, 0] = np.nan

X_init = torch.zeros((n, q))
X_inducing = torch.tensor(np.random.normal(size=(m, q))).float()
kernel = gp.kernels.Matern52(input_dim=q, lengthscale=torch.ones(q))

gp_module = gp.models.VariationalSparseGP(
    X=X_init,
    y=Y.T,
    kernel=kernel,
    Xu=X_inducing,
    likelihood=MaskedGaussian(),
    jitter=1e-4)

gplvm = gp.models.GPLVM(gp_module)
gplvm.autoguide('X', pyro.distributions.Delta)

optimizer = torch.optim.Adam(gplvm.parameters(), lr=0.01)
loss_fn = pyro.infer.TraceMeanField_ELBO().differentiable_loss

optimizer.zero_grad()
loss = loss_fn(gplvm.model, gplvm.guide)
torch_backward(loss)

for name, param in gplvm.named_parameters():
    print(name + ': ' + str(param.grad))

optimizer.step()

# X_map: tensor([[        nan,         nan],
#         [-8.4482e-04, -1.1908e-01],
#         [ 3.7835e-01, -1.4465e+00],
#         [-5.0512e-01,  4.7776e-01],
#         [-2.5080e+00,  4.8812e-01],
#         [ 8.2642e-01,  7.1484e-02],
#         [ 1.7435e+00, -1.3530e+00],
#         [ 2.1521e-01,  5.5909e-01],
#         [-2.0693e-01,  7.5668e-01],
#         [ 4.0503e-01,  1.9582e-01]])
# base_model.Xu: tensor([[nan, nan],
#         [nan, nan],
#         [nan, nan]])
# base_model.u_loc: tensor([[    nan,     nan,     nan],
#         [ 0.7670,  0.2054, -0.4713]])
# base_model.u_scale_tril_unconstrained:
# tensor([[[    nan,  0.0000,  0.0000],
#          [    nan,     nan,  0.0000],
#          [    nan,     nan,     nan]],

#         [[ 0.2107,  0.0000,  0.0000],
#          [-0.1242,  0.4675,  0.0000],
#          [ 0.1274, -0.2804,  0.9252]]])
# base_model.kernel.variance_unconstrained: tensor(nan)
# base_model.kernel.lengthscale_unconstrained: tensor([nan, nan])
# base_model.likelihood.variance_unconstrained: tensor(nan)
```

**Comment by fehiepsi** [comment]
@InfProbSciX I think you will need to also do `y = torch.masked_fill(y, y.isnan(), -999.)` if `y` has NaN value. [This note](https://github.com/tensorflow/probability/blob/master/discussion/where-nan.pdf) explains well why `NaN` happens during `backward`.

**Comment by InfProbSciX** [comment]
This worked! This is the class for reference:

```python
class MaskedGaussian(gp.likelihoods.Gaussian):
    def forward(self, f_loc, f_var, y=None):
        y_var = f_var + self.variance
        y_dist = dist.Normal(f_loc, y_var.sqrt())

        if y is not None:
            if y.isnan().any():
                y_dist = dist.MaskedDistribution(y_dist, ~y.isnan())
                y = torch.masked_fill(y, y.isnan(), -999.)

            y_dist = y_dist.expand_by(y.shape[:-f_loc.dim()]).to_event(y.dim())
        return pyro.sample(self._pyro_get_fullname("y"), y_dist, obs=y)
```

I think I'm slightly in favour of changing the existing `Gaussian` likelihood class actually, and I can add an optional argument to the `__init__` function like `mask_missing_data=False` as there's a lot of repeated code here. What would you say?

Also, another quick comment, would it be possible to use this likelihood with the `(Sparse)GPRegression` classes? I ask because `likelihood` doesn't seem to be an optional argument for those classes, so perhaps it'd be a bit more work to get them to work under the missing data regime.

**Comment by fehiepsi** [comment]
Yeah, you pointed it out, having a `mask` arg in the construction of Gaussian looks reasonable to me. For regression classes, I am not sure if there is an easy fix because internally, for prediction, we conditioned on `y`, instead of a latent `f`.
